### PR TITLE
feat(levm): opcodes gas consumption

### LINKED ISSUE #493: Opcode: GAS

### PR DESCRIPTION
**Pull request by juanimedone**
**Motivation**
Implement `GAS` opcode, which requires a first global handling of gas consumption.

**Description**
- Uses `VM::gas_limit` and `VM::consumed_gas` for handling all opcodes static and dynamic gas consumption.
- Check if gas is enough before executing each one of them.
- Save `warm_addresses` in `VM`.
- Fixes a `jump` bug and a `CALL` test.
- Adds gas checks to the tests. 
- `CALL`'s `code_execution_cost` is still to be implemented (#605). 

Closes #493 and #587.

### CODE REVIEW

File: crates/levm/src/vm.rs [hunk 1]
```diff
@@ -164,8 +213,12 @@ impl VM {
                     };
 
                     current_call_frame.stack.push(remainder);
+                    tx_env.consumed_gas += gas_cost::MOD
```
Comment (belfortep) [inline crates/levm/src/vm.rs:216]:
```suggestion
                    tx_env.consumed_gas += gas_cost::SMOD
```

File: crates/levm/src/vm.rs [hunk 2]
```diff
@@ -518,70 +728,148 @@ impl VM {
...
                 }
                 Opcode::MCOPY => {
-                    // spend_gas(3) + dynamic gas
                     let dest_offset = current_call_frame.stack.pop().unwrap().try_into().unwrap();
-                    let src_offset = current_call_frame.stack.pop().unwrap().try_into().unwrap();
-                    let size = current_call_frame.stack.pop().unwrap().try_into().unwrap();
+                    let src_offset: usize =
+                        current_call_frame.stack.pop().unwrap().try_into().unwrap();
+                    let size: usize = current_call_frame.stack.pop().unwrap().try_into().unwrap();
                     if size == 0 {
```
Comment (belfortep) [inline crates/levm/src/vm.rs:809]: Here if size is 0 I think we aren't consuming any gas, shouldn't we atleast consume `gas_cost::MCOPY_STATIC`?

File: crates/levm/tests/tests.rs [hunk 3]
```diff
@@ -1160,151 +1230,7 @@ fn mload_uninitialized_memory() {
 
     assert_eq!(loaded_value, U256::zero());
     assert_eq!(memory_size, U256::from(96));
-}
-
-#[test]
-fn pop_op() {
```
Comment (belfortep) [inline crates/levm/tests/tests.rs:1166]: Why this tests were deleted?

Comment (belfortep) [reply crates/levm/tests/tests.rs:1166]: nvm, they are at the end of the file 🫡

**Review by belfortep — COMMENTED** [review]
Great work! everything looks nice, just asked for some minor changes

File: crates/levm/src/memory.rs [hunk 4]
```diff
@@ -52,4 +52,17 @@ impl Memory {
...
     }
+
+    pub fn expansion_cost(&self, memory_byte_size: usize) -> u64 {
+        if memory_byte_size <= self.data.len() {
+            return 0;
+        }
+        let new_memory_size_word = (memory_byte_size as u64 + 31) / 32;
+        let new_memory_cost =
+            (new_memory_size_word * new_memory_size_word) / 512 + (3 * new_memory_size_word);
+        let last_memory_size_word = (self.data.len() as u64 + 31) / 32;
```
[the remark this answers is not in the document]
Comment (juanimedone) [reply crates/levm/src/memory.rs]: You're right, done!

File: crates/levm/src/vm.rs (hunk 2 above)
Comment (juanimedone) [reply crates/levm/src/vm.rs]: Yes, just moved that gas consumption before the `if` statement.

**Review by belfortep — APPROVED** [review]
LGTM! 🚀

**Review by jrchatruc — APPROVED** [review]

**Pull request merged**